influxql/ast.go

```diff
@@ -4199,6 +4350,148 @@ func EvalBool(expr Expr, m map[string]interface{}) bool {
 	return v
 }
 
+// TypeMapper maps a data type to the measurement and field.
+type TypeMapper interface {
+	MapType(measurement *Measurement, field string) DataType
+}
+
+type nilTypeMapper struct{}
+
+func (nilTypeMapper) MapType(*Measurement, string) DataType { return Unknown }
+
+// EvalType evaluates the expression's type.
+func EvalType(expr Expr, sources Sources, typmap TypeMapper) DataType {
+	if typmap == nil {
+		typmap = nilTypeMapper{}
+	}
+
+	switch expr := expr.(type) {
+	case *VarRef:
+		// If this variable already has an assigned type, just use that.
+		if expr.Type != Unknown && expr.Type != AnyField {
+			return expr.Type
+		}
+
+		var typ DataType
+		for _, src := range sources {
+			switch src := src.(type) {
+			case *Measurement:
+				t := typmap.MapType(src, expr.Val)
+				if typ == Unknown || t < typ {
+					typ = t
+				}
+			case *SubQuery:
+			FIELDS:
+				for _, f := range src.Statement.Fields {
+					if f.Name() == expr.Val {
+						t := EvalType(f.Expr, src.Statement.Sources, typmap)
+						if typ == Unknown || t < typ {
+							typ = t
+						}
+						break
+					} else if call, ok := f.Expr.(*Call); ok && (call.Name == "top" || call.Name == "bottom") {
+						// Handle tags selected in the top and bottom calls.
+						if len(call.Args) > 2 {
+							for _, arg := range call.Args[1 : len(call.Args)-1] {
+								if arg, ok := arg.(*VarRef); ok && arg.Val == expr.Val {
+									t := EvalType(arg, src.Statement.Sources, typmap)
+									if typ == Unknown || t < typ {
+										typ = t
+									}
+									break FIELDS
+								}
+							}
+						}
+					}
+				}
```

Comment:
It might be cleaner to break this block out to an `evalSubqueryType()` function. The indentation is pretty far over and it would allow you to remove the label on the `for`.